contracts/strategy/BaseStrategy.sol

```diff
@@ -0,0 +1,105 @@
+    using ERC165Query for address;
+
+    /// role allowed to invest/withdraw from yearn vault
+    bytes32 public constant MANAGER_ROLE = keccak256("MANAGER_ROLE");
+    /// role allowed to change settings such as max loss on withdraw from yearn vault
+    bytes32 public constant SETTINGS_ROLE = keccak256("SETTINGS_ROLE");
+    // underlying ERC20 token
+    IERC20 public underlying;
+    // the vault linked to this strategy
+    address public vault;
+
+    //
+    // Modifiers
+    //
+
+    modifier onlyManager() {
+        if (!hasRole(MANAGER_ROLE, msg.sender))
+            revert StrategyCallerNotManager();
+        _;
+    }
+
+    modifier onlySettings() {
+        if (!hasRole(SETTINGS_ROLE, msg.sender))
+            revert StrategyCallerNotSettings();
+        _;
+    }
+
+    modifier onlyAdmin() {
+        if (!hasRole(DEFAULT_ADMIN_ROLE, msg.sender))
+            revert StrategyCallerNotAdmin();
+        _;
+    }
+
+    /**
+     * Sets up the essential fields required for any strategy implementation.
+     *
+     * @notice This is ment to be called only from the #constructor of the implemeting contracts.
+     *
+     * @param _vault address of the vault that will use this strategy
+     * @param _underlying address of the underlying token
+     * @param _admin address of the administrator account
+     */
+    function setup(
+        address _vault,
+        address _underlying,
+        address _admin
+    ) internal {
+        if (_admin == address(0)) revert StrategyAdminCannotBe0Address();
+        if (_underlying == address(0))
+            revert StrategyUnderlyingCannotBe0Address();
+        if (!_vault.doesContractImplementInterface(type(IVault).interfaceId))
+            revert StrategyNotIVault();
+
+        vault = _vault;
+        underlying = IERC20(_underlying);
+
+        _setupRole(DEFAULT_ADMIN_ROLE, _admin);
+        _setupRole(MANAGER_ROLE, _vault);
+    }
+
+    /**
+     * Changes the account with the ADMIN role with the provided account address.
+     *
+     * @notice This is ment to be called only from the #transferAdminRights function of the implemeting contracts.
+     *
+     * @param _newAdmin The new admin account for the strategy.
+     */
+    function changeAdmin(address _newAdmin) internal {
+        if (_newAdmin == address(0)) revert StrategyAdminCannotBe0Address();
+        if (_newAdmin == msg.sender)
+            revert StrategyCannotTransferAdminRightsToSelf();
+
+        _setupRole(DEFAULT_ADMIN_ROLE, _newAdmin);
+
+        _revokeRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    /**
+     * Transfers administrator rights for the Strategy to another account,
+     * revoking current admin roles and setting up the roles for the new admin.
+     *
+     * @notice Can only be called by the account with the ADMIN role.
+     *
+     * @param _newAdmin The new ADMIN account.
+     */
+    function transferAdminRights(address _newAdmin) external virtual;
```

Comment:
What's the reason for this function not to be implemented?

Reply:
the reason is that strategies can have other roles associated with an admin account (ex YearnStrategy) and that roles should be also transferred to the new admin, so the idea was for strategies to provide an implementation for this and call `changeAdmin` internal function within that implementation. However, I think we can also provide a default implementation for this, as you might be suggesting here

Reply:
hummmm, I just wonder if it's worth having this common interface. You could just have the `transferAdminRights` public function and any strategy could either keep it or call it from another function right?

Reply:
that's one option. My intent was to retain the same interface meaning there is only one function addressing transferring of the admin rights. Check out this [commit](https://github.com/lindy-labs/sc_solidity-contracts/pull/181/commits/a84182b281ccffa84366264a94ec2135c02f2fb9)